Automattic/vip-go-node
# Preflight Checks: Add Docker build step

## Description

In order to have high fidelity preflight tests, this PR implements a new preflight check that tries to build a Docker image with a similar process that is used for the production image.

When Docker isn't available on the machine, it will warn the user, recommending to run the preflight tests on a machine with Docker available.

This is intended as a first iteration for a Docker check step, to pave the way for more complex and accurate Docker builds, that could, for instance, take in account the site environment variables, and Node version.

<strike>Currently this check tries to get the node version from a `.nvmrc` file. If this file doesn't exist, or doesn't contain a valid node version, it will fallback to the current version being used to run the preflight checks.</strike>

The Node.JS version is picked with a simple selector, using the [inquirer package](https://www.npmjs.com/package/inquirer). Additionally, to provide a non-interactive alternative, one can use the `--node-version <VERSION>` (or `-n <VERSION`) argument, where `<VERSION>` should be a major Node.JS version - 12, 14 or 16.

![image](https://user-images.githubusercontent.com/5804923/146575945-93d853b7-60eb-4637-bd63-d6fd69e95f92.png)

## Steps to Test

1. Check out PR.
1. Run `npm run build`.
1. `cd` to a VIP Go node site repository
2. Run `node [path-to-vip-go-node-repo]/preflight-checks/dist/index.js`
3. If `docker` is available (and the application builds), you should see the build step succeeding:
```
$ node ../vip-go-node/preflight-checks/dist/index.js
(...)
  [ Step 3/3 ] Building Docker image...
  [ Step 3/3 ] Step details: Trying to build a Docker image for the project, using a production-like build script
  Info: Found Docker Engine 20.10.12
  Info: Using Node.js 16.3.0 to build the image
  Info: Using GitHub repository organization/your-repository
  Info: Building Docker image...
  [ Step 3/3 ] Step successful 👍
```
